### PR TITLE
Move index dependent calculation to megacomplexes for speed-up

### PR DESCRIPTION
**Pull request by joernweissenborn**
This PR changes the way index dependent matrices are calculated. Instead of looping in the engine, the full matrix is now calculated by the megacomplex. This should speed up index dependent calculations.

~~It also fixed a bug in dampened oscillation, the matrix was initialized with ones instead zeros.~~

Note to reviewers: I needed to touch all megacomplexes. They all need (further) brush up, that will be done in follow-up PR.

### Change summary

- Change1
- Change2

### Checklist

- [x] ✔️ Passing the tests (mandatory for all PR's)
- [x] 🚧 Added changes to changelog (mandatory for all PR's)

### CODE REVIEW

**Review by jsnel — CHANGES_REQUESTED** [review]
1. examples are not running on CI, 
doas example is broken
locally the coherent artifact unit test is also broken (but can't see that in the CI yet)
(can't start proper review until that's fixed)

2. my terminal is spammed with deprecation warnings from numba:

```python
numba\core\ir_utils.py:2147: NumbaPendingDeprecationWarning: 
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'centers' of function '_calculate_coherent_artifact_matrix'.

For more information visit https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "..\pyglotaran\glotaran\builtin\megacomplexes\coherent_artifact\coherent_artifact_megacomplex.py", line 104:
@nb.jit(nopython=True, parallel=True)
def _calculate_coherent_artifact_matrix(
^

  warnings.warn(NumbaPendingDeprecationWarning(msg, loc=loc))
numba\core\ir_utils.py:2147: NumbaPendingDeprecationWarning: 
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'widths' of function '_calculate_coherent_artifact_matrix'.

For more information visit https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "..\pyglotaran\glotaran\builtin\megacomplexes\coherent_artifact\coherent_artifact_megacomplex.py", line 104:
@nb.jit(nopython=True, parallel=True)
def _calculate_coherent_artifact_matrix(
^

  warnings.warn(NumbaPendingDeprecationWarning(msg, loc=loc))
numba\core\ir_utils.py:2147: NumbaPendingDeprecationWarning: 
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'centers' of function '__numba_parfor_gufunc_0x1b19cd827a0'.

For more information visit https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "<string>", line 1:
<source missing, REPL/exec in use?>

  warnings.warn(NumbaPendingDeprecationWarning(msg, loc=loc))
numba\core\ir_utils.py:2147: NumbaPendingDeprecationWarning:
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'widths' of function '__numba_parfor_gufunc_0x1b19cd827a0'.
For more information visit https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "<string>", line 1:
<source missing, REPL/exec in use?>
```

File: benchmark/pytest/analysis/test_optimization_group.py [hunk 1]
```diff
@@ -44,7 +44,6 @@ class BenchmarkMegacomplex(Megacomplex):
     def calculate_matrix(
         self,
         dataset_model,
-        global_index: int | None,
```
Comment (joernweissenborn) [inline benchmark/pytest/analysis/test_optimization_group.py:47]: Note for myself: this is not enough

Comment (jsnel) [reply benchmark/pytest/analysis/test_optimization_group.py:47]: Ok, if you wish to resolve it another way in this PR, then make the commit to do so, or create a new issues to do it later and mark this comment as resolved.

**Comment by jsnel** [comment]
I just benchmarked one of the 'heaviest' example we have, pyglotaran_examples\test\simultaneous_analysis_6d_disp\sim_analysis_script_6d_disp.py and the results are impressive!

main: 32.1s
this PR: 18.2s 

Nice @joernweissenborn , I think @ism200 will happy to learn about this!

**Comment by s-weigand** [comment]
Even so [b7d0eeb](https://github.com/glotaran/pyglotaran/pull/1175/commits/b7d0eeb297303b3eeea7dd81fc724defb54f784f) didn't fix the strange CI issue IMHO we should keep it for consistency.

**Review by jsnel — DISMISSED** [review]
Reviewed and tested ok.

File: benchmark/pytest/analysis/test_optimization_group.py (hunk 1 above)
Comment (jsnel) [reply benchmark/pytest/analysis/test_optimization_group.py:47]: Please address in the subsequent PRs.

**Review by jsnel — APPROVED** [review]

**Pull request merged**